### PR TITLE
feat(rx): A new method of piping has been added

### PR DESCRIPTION
**Pull request by benlesh**
+ Adds `rx` a new method of piping observables. This allows any valid observable input to be passed as the first argument, and it will implicitly convert the argument to an Observable and pass it to the function that may be present in the second argument, the return value of THAT function is then passed an argument to the next function, and so on
+ Corrects the types of `Observable#pipe` such that anything can be returned at any step of the functional chain, however the first pipeable function must accept an `Observable<T>`.
+ Adds dtslint and runtime tests for `rx` and `pipe`.

NOTE: This does NOT deprecate`Observable#pipe`. That will be done in a follow up, and we need to define what timeline we'll take to remove that, as it's an API that is broadly used - could be v9, could be v10, could be never. At this point, this is alpha/beta functionality.

BREAKING CHANGE: `Observable#pipe` now allows any pipeable unary function as an argument, just so long as the types properly compose, this means in some cases it will now return `unknown` instead of `Observable<unknown>` the workaround is just to cast the result for those cases. (or you can break your operator piping into smaller pipe sets)

Addresses #7203 

### Note for later: Make sure to update squashed commit message to reflect `rx` and not `r`.

### CODE REVIEW

File: src/internal/util/r.ts
Comment (LcsGa) [inline src/internal/util/r.ts:1]: Won't the `operators[]` syntax be implemented?

**Comment by benlesh** [comment]
After some Core Team discussion, and some testing of the syntax, we've decided to move the implementation to where it takes an array of operators as the second argument. One of the example cases that compelled this change was this:

```ts
r(
  combineLatest([a, b]),
  map(([a, b]) => a + b),
  filter(n => n > 10),
)
.subscribe(console.log)
```

In the above scenario, muscle-memory for RxJS users leads them to believe that `combineLatest` is an operator, where it is in fact creating an observable.

The following was deemed to be more readable:

```ts
r(
  combineLatest([a, b]),
  [
    map(([a, b]) => a + b),
    filter(n => n > 10),
  ],
)
.subscribe(console.log)
```

### This PR is currently broken

I knowingly pushed it up this way so it can be addressed later. There's a problem with the TypeScript types when we do this. TypeScript can't seem to figure out how to type the operators as they go through the array.

**Comment by benlesh** [comment]
cc @DanielRosenwasser 🙏❤️

**Comment by jannispl** [comment]
Personally I am not a huge fan of the new array notation. One of my main issues when dealing with RxJS operators is getting the damn brackets right when refactoring code. Introducing new ones won't help.

**Comment by benlesh** [comment]
After speaking offline with the TypeScript folks, there's just no way to properly type the array notation. So this is going back to the more "flat" approach.

File: src/internal/util/r.ts
Comment (benlesh) [reply src/internal/util/r.ts:1]: Operators are just unary functions... meaning a functions that take one argument. Even in the current pipe functions in RxJS 6.x and 7.x, any unary function will work, it's just our types didn't support it. We're correcting that here.

File: spec/util/pipe-spec.ts [hunk 1]
```diff
@@ -31,3 +31,37 @@ describe('pipe', () => {
     expect(c(someObj)).to.equal(someObj);
   });
 });
+
+describe('r', () => {
```
Comment (cartant) [inline spec/util/pipe-spec.ts:35]: Nitpick:
```suggestion
describe('rx', () => {
```

File: src/internal/Observable.ts [hunk 2]
```diff
@@ -256,72 +255,79 @@ export class Observable<T> implements Subscribable<T> {
...
+    op1: UnaryFunction<Observable<T>, A>,
+    op2: UnaryFunction<A, B>,
+    op3: UnaryFunction<B, C>,
+    op4: UnaryFunction<C, D>,
+    op5: UnaryFunction<D, E>,
+    op6: UnaryFunction<E, F>,
+    op7: UnaryFunction<F, G>,
+    op8: UnaryFunction<G, H>,
+    op9: UnaryFunction<H, Observable<I>>,
     ...operations: OperatorFunction<any, any>[]
```
Comment (cartant) [inline src/internal/Observable.ts:317]: Can we replace the `any` usage here with `unknown`?

File: src/internal/Observable.ts [hunk 3]
```diff
@@ -256,72 +255,79 @@ export class Observable<T> implements Subscribable<T> {
...
+    op1: UnaryFunction<Observable<T>, A>,
+    op2: UnaryFunction<A, B>,
+    op3: UnaryFunction<B, C>,
+    op4: UnaryFunction<C, D>,
+    op5: UnaryFunction<D, E>,
+    op6: UnaryFunction<E, F>,
+    op7: UnaryFunction<F, G>,
+    op8: UnaryFunction<G, H>,
+    op9: UnaryFunction<H, I>,
+    ...operations: UnaryFunction<any, any>[]
```
Comment (cartant) [inline src/internal/Observable.ts:329]: Can we replace the `any` usage here with `unknown`?

File: src/internal/Observable.ts [hunk 4]
```diff
@@ -344,7 +350,7 @@ export class Observable<T> implements Subscribable<T> {
    *   .subscribe(x => console.log(x));
    * ```
    */
-  pipe(...operations: OperatorFunction<any, any>[]): Observable<any> {
+  pipe(...operations: UnaryFunction<any, any>[]): unknown {
```
Comment (cartant) [inline src/internal/Observable.ts:353]: Can we replace the `any` usage here with `unknown`?

File: src/internal/util/rx.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  source: ObservableInput<A>,
+  fn2: UnaryFunction<Observable<A>, B>,
+  fn3: UnaryFunction<B, C>,
+  fn4: UnaryFunction<C, D>,
+  fn5: UnaryFunction<D, E>,
+  fn6: UnaryFunction<E, F>,
+  fn7: UnaryFunction<F, G>,
+  fn8: UnaryFunction<G, H>,
+  fn9: UnaryFunction<H, I>,
+  ...fns: UnaryFunction<any, any>[]
```
Comment (cartant) [inline src/internal/util/rx.ts:70]: Same question as with `pipe`: Can we replace the `any` usage here with `unknown`?

**Review by cartant — APPROVED** [review]
LGTM, just a few questions about `any` usage and a nitpick.

File: src/internal/Observable.ts [hunk 6]
```diff
@@ -256,72 +255,79 @@ export class Observable<T> implements Subscribable<T> {
...
+    op3: UnaryFunction<B, C>,
+    op4: UnaryFunction<C, D>,
+    op5: UnaryFunction<D, E>,
+    op6: UnaryFunction<E, F>,
+    op7: UnaryFunction<F, G>,
+    op8: UnaryFunction<G, H>,
+    op9: UnaryFunction<H, Observable<I>>,
     ...operations: OperatorFunction<any, any>[]
   ): Observable<unknown>;
+  pipe<A, B, C, D, E, F, G, H, I>(
```
Comment (kolodny) [inline src/internal/Observable.ts:319]: I really wish we either had a dedicated pipe operator in js or at least had some way to use recursive/conditional types powerful enough to support not needing to do this.

File: src/internal/util/rx.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * This means it can be used for anything involving unary functions, just so long as the first unary function accepts an observable as its argument,
+ * and as long as the first argument to `r()` is a valid {@link ObservableInput}.
+ *
+ * This is the same as an ordinary functional {@link pipe}, except it has an implicit `from` as the second argument.
+ *
+ * The following are equivalent:
+ *
+ * ```ts
+ * r(source, map(x => x + 1), filter(x => x % 2 === 0));
```
Comment (kolodny) [inline src/internal/util/rx.ts:86]: Just need to update the comment to reflect the `r(source, [map(x => x + 1), filter(x => x % 2 === 0)]);` change

File: src/internal/util/rx.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ * pipe(from(source), map(x => x + 1), filter(x => x % 2 === 0));
+ * pipe(source, from, map(x => x + 1), filter(x => x % 2 === 0));
```
Comment (kolodny) [inline src/internal/util/rx.ts:88]: Probably also worth calling out that `rx([1, 2, 3]);` is an `Observable<number>` and not `Observable<number[]>` in the documentation (it's already in the type tests)

**Review by kolodny — APPROVED** [review]
🚀🚀🚀

File: src/internal/util/rx.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+ * This means it can be used for anything involving unary functions, just so long as the first unary function accepts an observable as its argument,
+ * and as long as the first argument to `r()` is a valid {@link ObservableInput}.
+ *
+ * This is the same as an ordinary functional {@link pipe}, except it has an implicit `from` as the second argument.
+ *
+ * The following are equivalent:
+ *
+ * ```ts
+ * r(source, map(x => x + 1), filter(x => x % 2 === 0));
+ * pipe(from(source), map(x => x + 1), filter(x => x % 2 === 0));
```
Comment (jakovljevic-mladen) [inline src/internal/util/rx.ts:87]: This doesn't seem to be valid. This does:
```suggestion
 * pipe(() => from(source), map(x => x + 1), filter(x => x % 2 === 0));
```

File: src/internal/util/rx.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * all functions have been called, and the result of the last function is returned.
+ *
+ * This means it can be used for anything involving unary functions, just so long as the first unary function accepts an observable as its argument,
+ * and as long as the first argument to `r()` is a valid {@link ObservableInput}.
+ *
+ * This is the same as an ordinary functional {@link pipe}, except it has an implicit `from` as the second argument.
+ *
+ * The following are equivalent:
+ *
+ * ```ts
```
Comment (jakovljevic-mladen) [inline src/internal/util/rx.ts:85]: Adding the source would help a lot. E.g.

```const source = Promise.resolve(1);```

or

```const source = [1, 2, 3];```

**Review by jakovljevic-mladen — CHANGES_REQUESTED** [review]
A lot of minor issues found: `r` to `rx` rename should be updated in couple of places, some generics issues need to be addressed and docs needs more info and some updates.

File: src/internal/util/rx.ts (hunk 9 above)
Comment (benlesh) [reply src/internal/util/rx.ts:87]: Fixed. It was a little different than what that said.

Comment (benlesh) [reply src/internal/util/rx.ts:87]: I fixed the types up a bit.

File: src/internal/util/rx.ts (hunk 10 above)
Comment (benlesh) [reply src/internal/util/rx.ts:85]: I added more context.

File: src/internal/util/rx.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+ * const promise = r(
+ *   of(1, 2, 3),
+ *   async (source) => {
+ *     const result = [];
+ *     await source.forEach(x => result.push(x));
+ *     return result;
+ *   },
+ * });
+ * ````
+ */
```
[the remark this answers is not in the document]
Comment (benlesh) [reply src/internal/util/rx.ts:109]: I fixed the types up, which should resolve that a little.

[the remark this answers is not in the document]
Comment (benlesh) [reply src/internal/util/rx.ts:109]: I also added `@param` and `@return` docs.

**Review by jakovljevic-mladen — APPROVED** [review]
LGTM overall, there are just some nitpicks that I'd like to be considered fixing.

**Pull request merged**